sycl/source/detail/graph_impl.hpp

```diff
@@ -51,13 +55,23 @@ class node_impl {
   /// use a raw \p this pointer, so the extra \Prev parameter is passed.
   void registerSuccessor(const std::shared_ptr<node_impl> &Node,
                          const std::shared_ptr<node_impl> &Prev) {
+    if (std::find(MSuccessors.begin(), MSuccessors.end(), Node) !=
+        MSuccessors.end()) {
+      return;
+    }
     MSuccessors.push_back(Node);
     Node->registerPredecessor(Prev);
   }
 
   /// Add predecessor to the node.
   /// @param Node Node to add as a predecessor.
   void registerPredecessor(const std::shared_ptr<node_impl> &Node) {
+    if (std::find_if(MPredecessors.begin(), MPredecessors.end(),
+                     [&Node](const std::weak_ptr<node_impl> &Ptr) {
+                       return Ptr.lock() == Node;
+                     }) != MPredecessors.end()) {
+      return;
```

Comment:
is it worth returning a `bool` here to report whether the node was successfully registered? same for `registerSuccessor`

Reply:
It could be as a "could be useful in the future" thing, I had added similar returns to adding/removing roots but then removed them as they turned out to not be needed in the end. If you think it's a useful addition I can add them, though I don't think there's any current use for them.